### ISSUE TITLE
Bug/feature in find_following_working_day

### ISSUE THREAD

**Issue by nedlowe** [novafloss/workalendar#194]
find_following_working_day returns the next non-weekend day, but not necessarily the next working day.

E.g. if a Sunday and a Monday are public holidays (assuming SAT/SUN weekend), and one wants to roll the Sunday to the next working day then it should return Tuesday.  Instead it returns Monday.

Since this function is often used for determining which days are holidays, it is non-trivial to just incorporate the holiday calendar as you get a chicken and egg situation (you can't check the holiday calendar when trying to calculate the holiday calendar).

**Comment by brunobord** [comment]
Mmmmm I see... you're right, this is a chicken & egg situation, here.
AFAICT, we're adding fixed holidays first, and then variable days. The "find_following_working_day" function should not be used for calculating holidays, but should rather only be used when the holiday calendar for this year is already computed.
I need to dig into it and try to find a case where this causes a bug.

**Comment by nedlowe** [comment]
The first two days of Lunar New Year 2013 were Sunday/Monday, so in (at least) Singapore and Hong Kong the holiday was Monday/Tuesday.

I have written a Singapore calendar with this case as a failing test.  I can submit it as a PR if you like.

**Comment by brunobord** [comment]
> I have written a Singapore calendar with this case as a failing test. I can submit it as a PR if you like.

Oh, that'd be awesome, please do!

**Comment by nedlowe** [comment]
Done.  Note that I know there is some stuff to cleanup - e.g. the super() calls double-populate some holidays - but the test case works as required.

**Comment by nedlowe** [comment]
Any thoughts on this Sir?  Many thanks.

**Comment by brunobord** [comment]
Hello,

Sorry about the lack of feedback, I've been a bit busy. I think I may start out of your PR/branch and rework it a bit before integrate it into the workalendar lib.
Thanks a lot for your time.

**Comment by brunobord** [comment]
@nedlowe quick and **important** question ; it looks like the Chinese New Year day in 2013 was on February the 10th. But on your branch, it's set on the 11th, and the shift day on the 12th:

```python
self.assertIn(date(2013, 2, 11), holidays)  # CNY
```

That's weird, because the 11th February 2013 falls on a Monday, so I think it's an error on your side, but just to make sure... Which one should I trust?

**Comment by brunobord** [comment]
I'm tempted to implement the following (pseudo-code):

```python
days = super(Singapore, self).get_variable_days(year)
pre_shifted_days = [
    (LunarCalendar.lunar(year, 1, 1), "Chinese New Year"),
    (LunarCalendar.lunar(year, 1, 2), "Chinese New Year"),
    (LunarCalendar.lunar(year, 4, 15), "Vesak Day")
    (date(year, 1, 1), "New Year"),
    (date(year, 5, 1), "Labour Day"),
    (date(year, 8, 9), "National Day")
]
# Singapore rolls all Sunday holidays.
# I feel this should be baked deeper in the class hierarchy
for holiday, label in pre_shifted_days:
    # Whether they're on Sunday or not, they can be added to the calendar
    days.append((holiday, label))
    if holiday.weekday() == SUN:
        days.append((holiday + timedelta(days=1), label + ' shift'))
```

this way, you'll have both "official" holidays, and shifted days in the variable days.

**Comment by nedlowe** [comment]
Hey no worries, was just checking in :-)

Here is the official Ministry of Manpower calendar for 2013: http://www.mom.gov.sg/newsroom/press-releases/2012/singapore-public-holidays-2013

So CNY was 10/11th.  The 10th was a Sunday so rolled to the next available date (12th).

The possible implementation doesn't solve this issue (unless I'm missing something) since the holiday on the 10th (Sunday) rolls to the 11th.

**Comment by brunobord** [comment]
> So CNY was 10/11th. The 10th was a Sunday so rolled to the next available date (12th).

oh! now I see. I guess that the right way to tackle this problem is to make a dedicated method in Singapore calendar. Will work on it as soon as I can

**Comment by nedlowe** [comment]
I'm happy to do it for Singapore, but I guess my point was this is probably
not a Singapore specific thing.  Same for Sunday rolls in general I guess...

**Comment by brunobord** [comment]
closed by #198

**Comment by brunobord** [comment]
[Singapore calendar available in the v1.3.0 release](https://github.com/novafloss/workalendar/releases/tag/1.3.0)
